src/baldr/edgeinfo.cc

```diff
@@ -122,28 +122,37 @@ EdgeInfo::GetNamesAndTypes(bool include_tagged_names) const {
 }
 
 // Get a list of tagged names
-std::vector<std::pair<std::string, uint8_t>> EdgeInfo::GetTaggedNamesAndTypes() const {
-  // Get each name
-  std::vector<std::pair<std::string, uint8_t>> name_type_pairs;
-  name_type_pairs.reserve(name_count());
-  const NameInfo* ni = name_info_list_;
-  for (uint32_t i = 0; i < name_count(); i++, ni++) {
-    // Skip any non tagged names
-    if (ni->tagged_) {
-      if (ni->name_offset_ < names_list_length_) {
-        std::string name = names_list_ + ni->name_offset_;
-        if (name.size() > 1) {
-          try {
-            auto num = static_cast<uint8_t>(name.at(0));
-            name_type_pairs.push_back({name.substr(1), num});
-          } catch (const std::logic_error&) { LOG_DEBUG("logic_error thrown for name: " + name); }
+const std::multimap<TaggedValue, std::string>& EdgeInfo::GetTags() const {
+  // we could check `tag_cache_.empty()` here, but many edges contain no tags
+  // and it would mean we traverse all names on each `GetTags` call
+  // for such edges
+  if (!tag_cache_ready_) {
```

Comment:
you could still use `tag_cache_` by setting the reservation/capacity to 0 on initialization and then resetting it here to at least 1 and then checking that as a sign for having tried to get the tag/values or not. it would be one less variable to track but its probably more confusing than just having the ready_ variable

Reply:
Since we are using `std::multimap` here it means we don't have "capacity" there. We probably could use `std::unordered_multimap` here(I had some reason to use ordered version here btw), which has `reserve` but it doesn't actually have `capacity` 😄 More than it I think reservation would mean additional dynamic memory allocation which we can avoid here. WDYT?

Reply:
ah i missed that detail. i think using the ready_ variable is the easiest to understand so i think we should stick with that. i assumed that we used the `unordered` version of this datastructure so i didnt even check it. in this case, there are likely to be, if any, only a tiny number of tag/value pairs so i dont think it matters performance-wise. at most you might have tunnel with a layer or bridge with a layer but probably no more than 2 on any real way in osm